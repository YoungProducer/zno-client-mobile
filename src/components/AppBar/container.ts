/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 2 March 2020
 *
 * Declare main types and interfaces for AppBar component.
 * Create function which connect actions
 * and/or variables from the redux-store to the component.
 */

/** External imports */
import { connect } from 'react-redux';
import { compose } from 'redux';
import { createStyles, Theme, withStyles, WithStyles } from '@material-ui/core/styles';
import { green } from '@material-ui/core/colors';

/** Application's imports */

/** Define Material ui styles */
const styles = (theme: Theme) =>
    createStyles({
        root: {
            background: green[700],
            width: '100%',
            height: 50,
        },
        container: {
            display: 'flex',
            justifyContent: 'center',
            alignItems: 'center',
            height: 'inherit',
        },
        navLink: {
            color: '#fff',
            fontSize: '1.2rem',
            marginRight: theme.spacing(3),
            '&:last-child': {
                marginRight: 0,
            },
        },
    });

/** Props which component get from the parent */
interface IOwnProps extends WithStyles<typeof styles>{}

/** Declare type which describe all props pushed to the component */
export type TAppBarProps =
    IOwnProps;

/** Export function to configure component via HOC */
export default compose(
    withStyles(styles, { withTheme: true }),
);
