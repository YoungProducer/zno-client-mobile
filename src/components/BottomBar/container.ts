/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 28 March 2020
 *
 * Container for BottomBar component.
 */

/** External imports */
import { connect } from 'react-redux';

/** Application's imports */
import { selectIsLoggedIn } from 'store/selectors/auth';
import { fetchLogoutAction } from 'store/actionCreators/auth';
import { RootState, setTestSuiteFinishedAction } from 'store/slices';

interface IStateProps {
    loggedIn: boolean;
}

interface IDispatchProps {
    logout: () => void;
    setTestSuiteFinished: (value: boolean) => void;
}

export type TBottomBarProps =
    IStateProps
    & IDispatchProps;

const mapStateToProps = (state: RootState): IStateProps => ({
    loggedIn: selectIsLoggedIn(state),
});

const mapDispatchToProps = (dispatch: any): IDispatchProps => ({
    logout: () => dispatch(fetchLogoutAction()),
    setTestSuiteFinished: (finished: boolean) =>
        dispatch(setTestSuiteFinishedAction(finished)),
});

export default connect<IStateProps, IDispatchProps>(
    mapStateToProps,
    mapDispatchToProps,
);
