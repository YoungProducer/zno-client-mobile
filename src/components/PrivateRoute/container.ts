/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 1 March 2020
 *
 * Define main types and interface related to PrivateRoute component.
 * Create function which connect variables from the redux store.
 */

/**External imports */
import { ReactNode } from 'react';
import { connect } from 'react-redux';
import { RouteProps } from 'react-router-dom';

// Applicaiton's imports
import { selectIsLoggedIn } from 'store/selectors/auth/signin';
import { RootState } from 'store/slices';

/**
 * Props which component get from the parent component.
 */
interface IOwnProps extends RouteProps {
    children: ReactNode;
}

/**
 * Props which component get from the redux store.
 */
interface IStateProps {
    isLoggedIn: boolean;
}

/**
 * Declare type which describe all props pushed to the component.
 */
export type TPrivateRoute =
    IOwnProps
    & IStateProps;

/**
 * Define function mapStateToProps.
 * Function select variables from the redux store.
 */
const mapStateToProps = (state: RootState): IStateProps => ({
    isLoggedIn: selectIsLoggedIn(state),
});

/**
 * Export function which connect actions
 * or/and variables from the redux store to component.
 */
export default connect<IStateProps, {}, IOwnProps>(
    mapStateToProps,
);
