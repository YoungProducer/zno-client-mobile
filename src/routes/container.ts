/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 29 February 2020
 *
 * Declare main interfaces and types for routes component.
 * Define function which connect actions
 * and/or variables from the redux-store to the component.
 */

/** External imports */
import { connect } from 'react-redux';

/** Application's imports */
import { selectIsLoggedIn } from 'store/selectors/auth';
import { RootState } from 'store/slices';

/** Props which component get from the parent */
interface IOwnProps {}

/** Props which component select from the redux-store */
interface IStateProps {
    isLoggedIn: boolean;
}

/** Props(actions) which component can dispatch */
interface IDispatchProps {}

/** Declare type which describe all props pushed to the component */
export type TRoutesProps =
    IOwnProps
    & IStateProps
    & IDispatchProps;

/** Create function which select variables from the redux-store and connect them to the component */
const mapStateToProps = (state: RootState): IStateProps => ({
    isLoggedIn: selectIsLoggedIn(state),
});

/** Export function which create configured HOC */
export default connect<IStateProps, IDispatchProps, IOwnProps>(
    mapStateToProps,
);
