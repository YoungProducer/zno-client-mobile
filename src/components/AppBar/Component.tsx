/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 2 March 2020
 *
 * Header component which contain main routes and another information or actions.
 */

/** External imports */
import React from 'react';
import CssBaseLine from '@material-ui/core/CssBaseline';
import Container from '@material-ui/core/Container';
import AppBar from '@material-ui/core/AppBar';

/** Application's imports */
import { TAppBarProps } from './container';
import NavigationLink from 'components/NavigationLink';

/** Create component */
const Component = (props: TAppBarProps) => {
    const { classes } = props;

    return (
        <React.Fragment>
            <CssBaseLine />
            <AppBar
                position='static'
                className={classes.root}
            >
                <Container
                    maxWidth='lg'
                    className={classes.container}
                >
                    <NavigationLink
                        navLink={{
                            to: '/',
                        }}
                        className={classes.navLink}
                        data-testid='home-navlink'
                    >
                        Головна
                    </NavigationLink>
                    <NavigationLink
                        navLink={{
                            to: '/subject-selection',
                        }}
                        className={classes.navLink}
                        data-testid='tests-navlink'
                    >
                        Тести
                    </NavigationLink>
                </Container>
            </AppBar>
        </React.Fragment>
    );
};

export default Component;
