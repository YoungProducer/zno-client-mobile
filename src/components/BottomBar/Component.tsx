/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 28 March 2020
 *
 * Component which display main navigation of app.
 */

/** External imports */
import React, { useState, useMemo } from 'react';
import { useHistory, useLocation } from 'react-router-dom';
import classNames from 'classnames';
import Collapse from '@material-ui/core/Collapse';
import Button from '@material-ui/core/Button';
import IconButton from '@material-ui/core/IconButton';
import ExpandLessIcon from '@material-ui/icons/ExpandLess';
import HomeIcon from '@material-ui/icons/Home';
import MoreHorizIcon from '@material-ui/icons/MoreHoriz';
import PersonIcon from '@material-ui/icons/Person';
import { makeStyles, createStyles, Theme } from '@material-ui/core';

/** Application's imports */
import { TBottomBarProps } from './container';

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        root: {
            position: 'absolute',
            minHeight: 70,
            borderRadius: 9,
            background: '#fff',
            filter: 'drop-shadow(0px 3px 3.5px rgba(0,0,0,0.16))',
            bottom: theme.spacing(1),
            left: theme.spacing(1),
            right: theme.spacing(1),
            transform: `translateY(0)`,
            transition: theme.transitions.create(['transform'], {
                duration: 200,
                easing: theme.transitions.easing.easeInOut,
            }),
        },
        rootHidden: {
            transform: `translateY(60%)`,
            transition: theme.transitions.create(['transform'], {
                duration: 200,
                easing: theme.transitions.easing.easeInOut,
            }),
        },
        innerContainer: {
            width: '100%',
            height: '100%',
        },
        actionsContainer: {
            width: '100%',
            height: '100%',
            minHeight: 70,
            display: 'flex',
            justifyContent: 'space-around',
            alignItems: 'center',
        },
        expandIconButton: {
            zIndex: 10,
            position: 'absolute',
            left: '50%',
            transform: `translate(-50%, -60%)`,
            borderRadius: '50%',
            background: '#fff',
            filter: 'drop-shadow(0px 3px 3.5px rgba(0,0,0,0.05))',
            transition: theme.transitions.create(['transform'], {
                duration: 200,
                easing: theme.transitions.easing.easeInOut,
            }),
            '&:hover': {
                background: '#fff',
            },
        },
        expandIconButtonHidden: {
            transform: `translate(-50%, -35%)`,
            transition: theme.transitions.create(['transform'], {
                duration: 200,
                easing: theme.transitions.easing.easeInOut,
            }),
        },
        expandIcon: {
            transform: `rotateX(180deg)`,
            transition: theme.transitions.create(['transform'], {
                duration: 200,
                easing: theme.transitions.easing.easeInOut,
            }),
        },
        expandIconHidden: {
            transform: `rotateX(0deg)`,
            transition: theme.transitions.create(['transform'], {
                duration: 200,
                easing: theme.transitions.easing.easeInOut,
            }),
        },
        actionIconButton: {
            background: '#fff',
            boxShadow: `0px 3px 3.5px rgba(0,0,0,0)`,
            transition: theme.transitions.create(['box-shadow'], {
                duration: 300,
                easing: theme.transitions.easing.easeInOut,
            }),
        },
        actionIconButtonActive: {
            boxShadow: `0px 3px 3.5px rgba(0,0,0,0.16)`,
            transition: theme.transitions.create(['box-shadow'], {
                duration: 500,
                easing: theme.transitions.easing.easeInOut,
            }),
        },
        actionIcon: {
            color: '#69deac',
            width: 35,
            height: 35,
        },
        expandActions: {
            padding: theme.spacing(1),
            display: 'flex',
            justifyContent: 'space-between',
        },
        button: {
            color: '#fff',
            background: '#656cff',
            borderRadius: 4,
            width: '45%',
        },
    }));

const Component = ({ loggedIn, logout }: TBottomBarProps) => {
    const classes = useStyles({});

    const history = useHistory();
    const location = useLocation();

    const [hidden, setHidden] = useState<boolean>(false);
    const [expand, setExpand] = useState<boolean>(false);

    const toggleHidden = () => {
        if (expand) {
            setExpand(false);
            window.setTimeout(() => {
                setHidden(!hidden);
            }, 300);
        } else {
            setHidden(!hidden);
        }
    };
    const toggleExpand = () => setExpand(!expand);

    const handleSignIn = () => history.push('/auth/signin');
    const handleSignUp = () => history.push('/auth/signup');
    const handleHome = () => history.push('/');
    const handleProfile = () => history.push('/user/profile');

    const activeIcons = useMemo(() => {
        const { pathname } = location;

        return {
            home: pathname === '/' || pathname === '/subject-selection',
            profile: pathname === '/user/profile',
        };
    }, [location]);

    return (
        <div
            className={classNames(classes.root, {
                [classes.rootHidden]: hidden,
            })}
        >
            <IconButton
                className={classNames(classes.expandIconButton, {
                    [classes.expandIconButtonHidden]: hidden,
                })}
                onClick={toggleHidden}
            >
                <ExpandLessIcon className={classNames(classes.expandIcon, {
                    [classes.expandIconHidden]: hidden,
                })}/>
            </IconButton>
            <div className={classes.innerContainer}>
                <div className={classes.actionsContainer}>
                    <IconButton
                        disabled={hidden}
                        className={classNames(classes.actionIconButton, {
                            [classes.actionIconButtonActive]: activeIcons.profile,
                        })}
                        onClick={handleProfile}
                    >
                        <PersonIcon className={classes.actionIcon}/>
                    </IconButton>
                    <IconButton
                        disabled={hidden}
                        className={classNames(classes.actionIconButton, {
                            [classes.actionIconButtonActive]: activeIcons.home,
                        })}
                        onClick={handleHome}
                    >
                        <HomeIcon className={classes.actionIcon}/>
                    </IconButton>
                    <IconButton
                        disabled={hidden}
                        className={classNames(classes.actionIconButton, {
                            [classes.actionIconButtonActive]: expand,
                        })}
                        onClick={toggleExpand}
                    >
                        <MoreHorizIcon className={classes.actionIcon}/>
                    </IconButton>
                </div>
                <Collapse in={expand} unmountOnExit>
                    <div className={classes.expandActions}>
                        { loggedIn && (
                            <Button
                                className={classes.button}
                                variant='contained'
                                disableElevation
                                onClick={logout}
                            >
                                Вийти
                            </Button>
                        )}
                        { !loggedIn && (
                            <>
                                <Button
                                    className={classes.button}
                                    variant='contained'
                                    disableElevation
                                    onClick={handleSignIn}
                                >
                                    Увійти
                                </Button>
                                <Button
                                    className={classes.button}
                                    variant='contained'
                                    disableElevation
                                    onClick={handleSignUp}
                                >
                                    Зареєструватися
                                </Button>
                            </>
                        )}
                    </div>
                </Collapse>
            </div>
        </div>
    );
};

export default Component;
