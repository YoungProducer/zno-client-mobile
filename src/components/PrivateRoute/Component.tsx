/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 1 March 2020
 *
 * Permission route which automaticaly redirect from protected pages if user is not logged in to SignIn page.
 */

/** External imports */
import React from 'react';
import { Redirect, Route } from 'react-router-dom';

/** Application's imports */
import { TPrivateRoute } from './container';

/** Create component */
const Component = ({ isLoggedIn, children, ...rest }: TPrivateRoute) => (
    <Route
        {...rest}
        render={({ location }) => {
            const pathname = location
                ? `/auth/signin?next=${location.pathname}${location.search}`
                : '/auth/signin';

            return isLoggedIn
                ? children
                : <Redirect
                    to={{
                        pathname: '/auth/signin',
                        state: { from: location },
                    }}
                />;
        }
        }
    />
);

export default Component;
