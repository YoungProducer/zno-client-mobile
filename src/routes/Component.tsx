/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 29 February 2020
 *
 * Component which contain main routes of application.
 */

/** External imports */
import React, { Suspense } from 'react';
import { Router, Route, Redirect, Switch } from 'react-router-dom';

/** Application's imports */
import PrivateRoute from 'components/PrivateRoute';
import { TRoutesProps } from './container';
import FullPageLoading from 'components/custom/FullPageLoading';
import BottomBar from 'components/BottomBar';

/** Lazy loaded components */
const SignIn = React.lazy(() => import('components/SignIn'));
const SignUp = React.lazy(() => import('components/SignUp'));
const SubjectSelection = React.lazy(() => import('./SubjectSelection'));
const SubjectConfiguration = React.lazy(() => import('./SubjectConfiguration'));
const TestSuite = React.lazy(() => import('./TestSuite'));

/** Create component */
const Component = ({ isLoggedIn }: TRoutesProps) => {
    return (
        <Suspense fallback={<FullPageLoading />}>
            <BottomBar />
            <Switch>
                <Route exact path='/'>
                    <Redirect to={{
                        pathname: '/subject-selection',
                        state: { from: '/' },
                    }}/>
                </Route>
                <Route exact path='/subject-selection'>
                    <SubjectSelection />
                </Route>
                <PrivateRoute
                    exact
                    path='/subject-selection/subject-configuration'
                >
                    <SubjectConfiguration />
                </PrivateRoute>
                <PrivateRoute
                    exact
                    path='/subject-configuration/:subjectId'
                >
                    <SubjectConfiguration />
                </PrivateRoute>
                <PrivateRoute
                    exact
                    path='/test-suite'
                >
                    <TestSuite />
                </PrivateRoute>
                <Route exact path='/auth/signin' component={SignIn} />
                <Route exact path='/auth/signup'>
                    { isLoggedIn
                        ? <Redirect
                            to={{
                                pathname: '/subject-selection',
                                state: { from: '/auth/signup' },
                            }}
                        />
                        : <SignUp />
                    }
                </Route>
            </Switch>
        </Suspense>
    );
};

export default Component;
