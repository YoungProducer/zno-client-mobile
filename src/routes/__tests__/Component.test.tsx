/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 29 February 2020
 *
 * Create test suites for routes component.
 * @jest-environment jsdom
 */

/** External imports */
import React from 'react';
import { render } from 'react-dom';
import { MemoryRouter } from 'react-router-dom';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

/** Application's imports */
import Component from '../Component';
import { TRoutesProps } from '../container';
import { RootState } from 'store/slices';
import store from 'store/__mocks__/mockedStore';

describe('Routes component', () => {
    /** Define required props */
    const requiredProps: TRoutesProps = {
        isLoggedIn: false,
    };

    test('Is match snapshot', () => {
        /** Render component in shallow */
        const root = shallow(<Component {...requiredProps}/>);

        /** Assert component matches snapshot */
        expect(root).toMatchSnapshot();
    });
});
