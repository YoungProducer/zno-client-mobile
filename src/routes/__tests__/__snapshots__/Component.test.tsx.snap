// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Routes component Is match snapshot 1`] = `
<Switch>
  <Route
    exact={true}
    path="/"
  >
    <Redirect
      to={
        Object {
          "pathname": "/subject-selection",
          "state": Object {
            "from": "/",
          },
        }
      }
    />
  </Route>
  <Route
    exact={true}
    path="/subject-selection"
  >
    <Component />
  </Route>
  <Connect(Component)
    exact={true}
    path="/subject-selection/subject-configuration"
  >
    <Component />
  </Connect(Component)>
  <Connect(Component)
    exact={true}
    path="/subject-configuration/:subjectId"
  >
    <Component />
  </Connect(Component)>
  <Connect(Component)
    exact={true}
    path="/test-suite"
  >
    <Component />
  </Connect(Component)>
  <Route
    component={[Function]}
    exact={true}
    path="/auth/signin"
  />
  <Route
    exact={true}
    path="/auth/signup"
  >
    <Connect(Component) />
  </Route>
</Switch>
`;
